sdk/src/utils/graphs/adjacency-pool-graph.ts

```diff
@@ -0,0 +1,195 @@
+      return [AddressUtil.toString(startMint), AddressUtil.toString(endMint)] as const;
+    });
+    const walkMap = findWalks(
+      searchTokenPairsAddrs,
+      this.graph,
+      options?.intermediateTokens.map((token) => AddressUtil.toString(token))
+    );
+
+    const results = searchTokenPairsAddrs.map(([startMint, endMint]) => {
+      const searchRouteId = PoolGraphUtils.getSearchRouteId(startMint, endMint);
+      const internalRouteId = getInternalRouteId(startMint, endMint);
+      const routesForSearchPair = walkMap[internalRouteId];
+      const paths = routesForSearchPair ? routesForSearchPair.map<Route>((route) => {
+        return {
+          startTokenMint: startMint,
+          endTokenMint: endMint,
+          hops: getHopsFromRoute(internalRouteId, searchRouteId, route),
+        };
+      }) : [];
+
+
+      return [searchRouteId, paths] as const;
+    });
+
+    return results;
+  }
+}
+
+
+
+function getHopsFromRoute(internalRouteId: string, searchRouteId: string, route: string[]): Hop[] {
+  const [intStartA] = PoolGraphUtils.deconstructRouteId(internalRouteId);
+  const [searchStartA] = PoolGraphUtils.deconstructRouteId(searchRouteId);
+  const shouldReverseRoute = searchStartA !== intStartA;
+  const finalRoutes = shouldReverseRoute ? route.reverse() : route;
+
+  return finalRoutes.map(hopStr => { return { poolAddress: new PublicKey(hopStr) } })
+}
+
+
+type AdjacencyPoolGraphMap = Record<string, Array<PoolGraphEdge>>;
+
+type PoolGraphEdge = {
+  address: string;
+  otherToken: string;
+};
+
+// A record of route-id (tokenA-tokenB) to a list of edges
+type PoolWalks = Record<string, string[][]>;
+
+function buildPoolGraph(pools: PoolTokenPair[]) {
+  const poolGraphSet = pools.reduce((poolGraph: Record<string, PoolGraphEdge[]>, pool) => {
+    const { address, tokenMintA, tokenMintB } = pool;
+    const [addr, mintA, mintB] = AddressUtil.toStrings([address, tokenMintA, tokenMintB]);
+
+    if (poolGraph[mintA] === undefined) {
+      poolGraph[mintA] = [];
+    }
+
+    if (poolGraph[mintB] === undefined) {
+      poolGraph[mintB] = [];
+    }
+
+    const existingAddressesForA = poolGraph[mintA].map((p) => p.address);
+    const existingAddressesForB = poolGraph[mintB].map((p) => p.address);
+
+
+    if (!existingAddressesForA.includes(addr)) {
+      poolGraph[mintA].push({ address: addr, otherToken: mintB });
+    }
+
+    if (!existingAddressesForB.includes(addr)) {
+      poolGraph[mintB].push({ address: addr, otherToken: mintA });
+    }
+
+    return poolGraph;
+  }, {});
+
+  return Object.fromEntries(
+    Object.entries(poolGraphSet).map(([mint, otherMints]) => [mint, Array.from(otherMints)])
+  );
+}
+
+// This is currently hardcoded to find walks of max length 2, generalizing to longer walks
+// may mean that a adjacency matrix might have better performance
+function findWalks(
+  tokenPairs: (readonly [string, string])[],
```

Comment:
I noticed that there are edge case inputs.
e.g.: [SOL, SOL], [USDC, USDC]

We didn't think of it because we are thinking of it for trading purposes, but it is likely to be the first input for users who use it to try arbitrage.
It might be better to handle this independently from the current logic.🤔

- single hop: no routes
- two hops: same pair with different fee tiers
  ok: SOL -- SOL/USDC(64) --> USDC -- SOL/USDC(8) --> SOL
  ok: SOL -- SOL/USDC(8) --> USDC --> SOL/USDC(64) --> SOL
  ng: SOL -- SOL/USDC8(64) --> USDC -- SOL/USDC(64) --> SOL

Reply:
ah yes. given that we don't use it, i would say let's throw an error and mark unsupported. Devs who has a need for this usecase can contribute themselves since this is an open-source project.

Reply:
to clarify, i ended up not throwing an error and these routes will just return [].